### PR TITLE
parser: support the syntax of "CharsetKw CharsetName OptBinMod"

### PR DESCRIPTION
**Pull request by zimulala**
Before this PR:
```
tidb>CREATE TABLE foo (name CHAR(50) CHARACTER SET utf8 BINARY);
ERROR 1105 (HY000): line 1 column 57 near ")" (total length 58)
tidb>ALTER TABLE foo CHANGE COLUMN name b varchar(255) CHARACTER SET utf8 BINARY;
ERROR 1105 (HY000): line 1 column 75 near "" (total length 75)
```

### CODE REVIEW

File: ast/ast.go [hunk 1]
```diff
@@ -83,6 +83,12 @@ type ExprNode interface {
 	Format(w io.Writer)
 }
 
+// OptBinary is used for parser.
+type OptBinary struct {
```
Comment (XuHuaiyu) [inline ast/ast.go]: s/ OptBinary/ OptBinaryCharset ?

File: parser/parser.y [hunk 2]
```diff
@@ -782,6 +782,7 @@ import (
 	FloatOpt		"Floating-point type option"
 	Precision		"Floating-point precision option"
 	OptBinary		"Optional BINARY"
+	OptBinMod		"Optional BINARY mode"
```
Comment (jackysp) [inline parser/parser.y]: I think 'mod' sometime has another mean. This word is not so long, there is no need to omit the last 'e'.

File: ast/ast.go (hunk 1 above)
Comment (zimulala) [reply ast/ast.go]: I refer to MySQL.

File: parser/parser.y (hunk 2 above)
Comment (zimulala) [reply parser/parser.y]: I refer to MySQL.

Comment (jackysp) [reply parser/parser.y]: ok

**Comment by coocood** [comment]
LGTM

**Pull request merged**